### PR TITLE
feat: Add Service field to Rollout Experiment to allow service creation

### PR DESCRIPTION
**Pull request by daniddelrio**
Follow-up to: #2624 

#2624 rolled back the change to allow rollout experiments with no weights to still create services. However, we still want the allow the users to choose to create services for their experiments even with no weight/traffic routing enabled. This PR makes it so that users can set the Service attribute for their rollout experiments (with an optional `name` field), giving them the ability to still create services without setting `weight`.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/) with a list of types and scopes found [here](https://github.com/argoproj/argo-rollouts/blob/master/.github/workflows/pr-title-check.yml), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [ ] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

File: rollout/experiment.go [hunk 1]
```diff
@@ -64,8 +64,11 @@ func GetExperimentFromTemplate(r *v1alpha1.Rollout, stableRS, newRS *appsv1.Repl
 			Name:     templateStep.Name,
 			Replicas: templateStep.Replicas,
 		}
-		if templateStep.Weight != nil {
+		if templateStep.Weight != nil || templateStep.Service != nil {
 			template.Service = &v1alpha1.TemplateService{}
+			if templateStep.Service != nil && templateStep.Service.Name != "" {
```
Comment (alexef) [inline rollout/experiment.go:69]:
```suggestion
			if templateStep.Service.Name != "" {
```

you already checked for service not being nil

File: rollout/experiment_test.go [hunk 2]
```diff
@@ -820,3 +820,48 @@ func TestRolloutCreateExperimentWithService(t *testing.T) {
...
+	steps := []v1alpha1.CanaryStep{{
+		Experiment: &v1alpha1.RolloutExperimentStep{
+			Templates: []v1alpha1.RolloutExperimentTemplate{
+				// Service should be created for "stable-weightless-template"
+				{
+					Name:     "stable-weightless-template",
+					SpecRef:  v1alpha1.StableSpecRef,
+					Replicas: pointer.Int32Ptr(1),
+					Service: &v1alpha1.TemplateService{
+						Name: "test-service",
```
Comment (alexef) [inline rollout/experiment_test.go:836]: please add test without name (if it exists, please ignore comment)

**Review by alexef — CHANGES_REQUESTED** [review]
hmm, shouldn't you pass template.Service.Name in CreateService call? instead of rs.Name 

https://github.com/daniddelrio/argo-rollouts/blob/service-in-experiment/experiments/experiment.go#L304

File: rollout/trafficrouting_test.go [hunk 3]
```diff
@@ -248,6 +248,7 @@ func TestRolloutWithExperimentStep(t *testing.T) {
 					SpecRef:  "canary",
 					Replicas: pointer.Int32Ptr(1),
 					Weight:   pointer.Int32Ptr(5),
+					Service:  &v1alpha1.TemplateService{},
```
Comment (zachaller) [inline rollout/trafficrouting_test.go:251]: Why did this have to be added?

File: rollout/experiment.go (hunk 1 above)
Comment (daniddelrio) [reply rollout/experiment.go:69]: This `templateStep.Service != nil` is for the case where templateStep.Weight != nil and templateStep.Service == nil, else `templateStep.Service.Name` will return an invalid memory address error

File: rollout/trafficrouting_test.go (hunk 3 above)
Comment (daniddelrio) [reply rollout/trafficrouting_test.go:251]: Actually that didn't need to be. Thanks

File: rollout/experiment.go (hunk 1 above)
Comment (alexef) [reply rollout/experiment.go:69]: but you already check it on line 67, so it can never be nil on line 69, right?

Comment (daniddelrio) [reply rollout/experiment.go:69]: Line 67 will pass even if Service is nil as long as Weight is not nil. Confirmed it by rerunning the unit tests as well

File: rollout/experiment_test.go (hunk 2 above)
Comment (daniddelrio) [reply rollout/experiment_test.go:836]: Done. See: https://github.com/argoproj/argo-rollouts/pull/2633/files#diff-4ee6ea6eacb6c4ad54a87a51cdba981669b25bbd2d30a0b51bbd8fcd56bce155R870

**Review by alexef — APPROVED** [review]

**Review by zachaller — APPROVED** [review]

**Pull request merged**